Start work on removing .staging directory

anikitenko opened the pull request:
This PR created to change the way ".staging" directory was used.
In this PR:
* temp directory is determined based on OS you are running
* directory in "temp" one is named according to cons `OpsDir`
* file that contains information about contents in that directory is const `TMPContentsFile`
* another temp directories are created using Go `ioutil.TempDir` with prefix const `TMPStagingDirectoryPrefix`
* Uniqueness is done using the following methods:
  * If user uses `run` command then full path to it's binary is used (ex. `/Users/alexvwan/dev/src/test/server`) with it's association to new unique temp directory
  * If user uses `load` command then full path to current directory + gathered from manifest `manifest.Program`

anikitenko commented:
So, in other words, instead of using ".staging" in each directory, now everything is located at: `<system temp dir>/<OpsDir>/*`. And information of existence of anything is located at `<system temp dir>/<OpsDir>/<TMPContentsFile>`

anikitenko commented:
BTW, such logic of defining uniqueness is only valid until user don't specify parameters related to location of boot images.. I thought that if user specify any than responsibility exactly on users's side

eyberg commented:
is there any reason besides debugging purposes that we even want to keep a .staging directory @tijoytom  && @mkhon ?? like - once the image is built imo we should just nuke it right? keep the resulting image but not the tmp stuff? and we could store the built image into something like ~/.ops/projects/my.img or something like that to help facilitate:

a) running it from anywhere (eg: outside the project root)

b) having a known good location to drop them in

also -- it doesn't need to be addressed in this pr but imo https://github.com/nanovms/ops/issues/196 prob. should go in the .staging (whatever becomes of this)

mkhon commented:
Can we keep mkfs, boot.img and stage3.img in a fixed location (e.g.
somewhere under ~/.ops)?

On Fri, 15 Feb 2019 at 06:48, Ian Eyberg <notifications@github.com> wrote:

> is there any reason besides debugging purposes that we even want to keep a
> .staging directory @tijoytom <https://github.com/tijoytom> && @mkhon
> <https://github.com/mkhon> ?? like - once the image is built imo we
> should just nuke it right? keep the resulting image but not the tmp stuff?
> and we could store the built image into something like
> ~/.ops/projects/my.img or something like that to help facilitate:
>
> a) running it from anywhere (eg: outside the project root)
>
> b) having a known good location to drop them in
>
> also -- it doesn't need to be addressed in this pr but imo #196
> <https://github.com/nanovms/ops/issues/196> prob. should go in the
> .staging (whatever becomes of this)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nanovms/ops/pull/208#issuecomment-463849476>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC4fnR9mmS9z2CG9fJCuqK5LWP6agfOZks5vNfXHgaJpZM4a8avZ>
> .
>

eyberg commented:
Id strongly prefer that as it can be reused

anikitenko commented:
@mkhon I have open PR to keep mkfs, boot.img and stage3.img in a fixed location #195 but work in progress there..

tijoytom commented on lepton/const.go:
```diff
@@ -21,26 +21,34 @@ const manifest string = `(
 const PackageBaseURL string = "https://storage.googleapis.com/packagehub/%v"
 const PackageManifestURL string = "https://storage.googleapis.com/packagehub/manifest.json"
 const PackageManifestFileName string = "manifest.json"
+const PackageRootPath string = "sysroot"
```
This should not be a global const if it's used only in one function.

anikitenko replied:
Hm.. I thought it's better to move "sysroot" to const instead of hardcoding. If you think it's not needed, I'll remove it

tijoytom replied:
I am fine if you want to keep it a const but then don't export it.(ie uppercase P)

anikitenko replied:
ah, okay, I'll make it not exported

tijoytom commented on lepton/image.go:
```diff
@@ -74,23 +74,36 @@ func addHostName(m *Manifest, c *Config) {
 }
 
 func addFilesFromPackage(packagepath string, m *Manifest) {
```
I don't think these changes work, i might be mistaken. Also there is no point in traversing the directory twice.

anikitenko replied:
I tested these changes and they are working. Also I'm not traversing directory twice: first `filepath.Walk` traversing "sysroot", second one is it's parent directory ignoring "sysroot".

tijoytom commented on lepton/config.go:
```diff
@@ -19,6 +19,7 @@ type Config struct {
 	RunConfig    RunConfig
 	Force        bool
 	TargetRoot   string
+	LocalTMPDir  string
```
TMP  should not be a config.

anikitenko replied:
Okay, I'll remove this from config

tijoytom commented on lepton/image.go:
```diff
@@ -204,17 +214,26 @@ func addMappedFiles(src string, dest string, m *Manifest) {
 	// invoke mkfs to create the filesystem ie kernel + elf image
-	createFile(mergedImg)
-	args := []string{}
+	mergedImgPath := filepath.Join(c.LocalTMPDir, mergedImg)
```
Why is this necessary? creating and closing the temp file. Any way it's going to be created.

anikitenko replied:
I thought that there is no need to keep it open...

anikitenko replied:
ah, got what you mean.. I'll remove `mergedImg.Close()`

tijoytom commented on lepton/image.go:
```diff
@@ -230,19 +249,35 @@ func buildImage(c *Config, m *Manifest) error {
 		return errors.Wrap(err, 1)
 	}
 
+	bootImageOpen, err := os.Open(c.Boot)
```
Again why do you want to open this file?

tijoytom commented on lepton/image.go:
```diff
@@ -230,19 +249,35 @@ func buildImage(c *Config, m *Manifest) error {
+	}
+	defer bootImageOpen.Close()
+
+	mergedImgPathOpen, err := os.Open(mergedImgPath)
```
same as above

anikitenko replied:
`c.Boot` and `mergedImgPath` are need to be open to perform `io.Copy` to final image. I thought it's better to rewrite code to use `io.Copy` instead of running shell command `cat`

tijoytom replied:
is there a problem with cat? if not  i would keep it as such, using io.Copy buys you nothing, it introduce more code while solving a non-existing problem.

anikitenko replied:
Actually there is no problem with `cat` but `io.Copy` is more a "Go way" of working with binary data..

mkhon replied:
I would prefer io.Copy as dependencies on external programs are bad

eyberg replied:
yeh - it might be a bit slower but agreed - we should have as little exec.command as possible

anikitenko commented:
I made changes based on discussions. I noticed that "image" is now placed in "$HOME/.ops/images" but it's name is hardcoded to "image". But what will happen if I run 2 images? I thought it's better to place it in temp directory determined by `GetTMPPathByProgramPath` func.. but for now I'm not exactly sure on how to do this.. :(

anikitenko commented:
Just now an idea came to my mind.. when user has "shared" account on some server, some administrators can restrict access to `/tmp` directory.. What do you think if I rewrite `GetTMPPathByProgramPath` func to set path to staging like the following: `filepath.Join(GetOpsHome(), "staging")`? And the other functionality of func will remain? I think that this could resolve problem #196 and will implement in #210

eyberg commented:
the tmp. image we don't want to keep around for anything other than debug purposes but for built images we need to store somewhere so it can be re-used again - in a known place - the images do need to be named in order for this to work but that wasn't clearly communicated - i suppose for now we can allow the old behavior by naming the image by default whatever directory it is in like go does but there should be a name argument && || spot in the config.json that will name it properly

sorry for the confusion

anikitenko commented:
Ah, I see... work on that could be next step from moving from ".staging" directory.

anikitenko commented:
Updated pull request to include fix of #217

anikitenko commented:
Also in these changes I changed `os.TempDir` to `GetOpsHome()` in order to make sure that errors like #196 won't happen

And also in these changes I added funcs `RemoveTMPPathByProgramPath` and `RemoveImagePathByProgramPath` so they can be used to implement #147

tijoytom commented on lepton/image.go:
```diff
@@ -225,19 +241,43 @@ func buildImage(c *Config, m *Manifest) error {
+	}
+
+	if len(c.Debugflags) == 0 {
+		if err := RemoveTMPPathByProgramPath(programAbsPath); err != nil {
```
Not sure i understand this, how is removing temp related to debug flags?

anikitenko replied:
Ah, also forgot to mention that I added the following:
* If we have debug option specified than tempimage will remain
* If nothing debug specified that tempimage is removed along with it's temp directory at `$HOME/.ops/<temp_path>`

anikitenko replied:
As @eyberg said that `tempimage` is more for debug purposes...

tijoytom replied:
Sorry for the confusion, tempimage is an artifact of how we build the images. First we merge kernel an user executable to a file (which is the tempimage), then we merge tempimage and the boot loader to final image. So you can remove the temp-image right after the final merge.

anikitenko replied:
I see.. but again, for debugging purposes maybe we can leave it if user specifies debug option?

eyberg replied:
i don't see a need for tempimage or any of the other artifacts to be present -- '-d' works fine w/the resulting built image - for instance in the old model -- 'image' we want to keep but anything in ~/.staging we don't want to keep

anikitenko replied:
Actually tempimage is not stored in .staging, everything is under ~/.ops, but Only in case someone specified -d option, tempimage is saved, in other cases it’s removed

eyberg replied:
tempimage does not need to be saved at all -- '-d' only uses the built image - we can completely remove tempimage

anikitenko replied:
Ah, okay, I’ll update PR to always remove it

anikitenko replied:
Updated PR to always remove tempimage

tijoytom commented on lepton/images.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	// Read contents of file which describes contents of images data
+	contentJSON, err := ioutil.ReadFile(imagesOpsFilePath)
```
Not sure i understand why all this is necessary, are you storing the list of temp files in a json and using it to delete them later? Why can't you just delete the temp folder ie `os.RemoveAll(temp)` would delete everything.

anikitenko commented:
Closing because PR is outdated
